node/network/collator-protocol/src/lib.rs

```diff
@@ -43,124 +43,285 @@ const LOG_TARGET: &'static str = "parachain::collator-protocol";
-		target: LOG_TARGET,
-		rep = ?rep,
-		peer_id = %peer,
-		"reputation change for peer",
-	);
-
-	ctx.send_message(AllMessages::NetworkBridge(
-		NetworkBridgeMessage::ReportPeer(peer, rep),
-	)).await;
+    tracing::trace!(
+        target: LOG_TARGET,
+        rep = ?rep,
+        peer_id = %peer,
+        "reputation change for peer",
+    );
+
+    ctx.send_message(AllMessages::NetworkBridge(
+        NetworkBridgeMessage::ReportPeer(peer, rep),
+    ))
+    .await;
+}
+
+pub(crate) mod fitness_metrics {
+    use super::{modify_reputation, CollatorId, CollatorProtocolMessage, PeerId, SubsystemContext};
+    use polkadot_node_network_protocol::UnifiedReputationChange as Rep;
+
+    pub(crate) const CACHE_SIZE: usize = 100;
+
+    pub(crate) const COST_UNEXPECTED_MESSAGE: Rep = Rep::CostMinor("An unexpected message");
+    /// Message could not be decoded properly.
+    pub(crate) const COST_CORRUPTED_MESSAGE: Rep = Rep::CostMinor("Message was corrupt");
+    /// Network errors that originated at the remote host should have same cost as timeout.
+    pub(crate) const COST_NETWORK_ERROR: Rep = Rep::CostMinor("Some network error");
+    pub(crate) const COST_REQUEST_TIMED_OUT: Rep =
+        Rep::CostMinor("A collation request has timed out");
+    pub(crate) const COST_INVALID_SIGNATURE: Rep =
+        Rep::Malicious("Invalid network message signature");
+    pub(crate) const COST_WRONG_PARA: Rep =
+        Rep::Malicious("A collator provided a collation for the wrong para");
+    pub(crate) const COST_UNNEEDED_COLLATOR: Rep = Rep::CostMinor("An unneeded collator connected");
+
+    pub(crate) const COST_REPORT_BAD: Rep =
+        Rep::Malicious("A collator was reported by another subsystem");
+    pub(crate) const BENEFIT_NOTIFY_GOOD: Rep =
+        Rep::BenefitMinor("A collator was noted good by another subsystem");
+
+    const COLLATOR_METRICS: [Rep; 9] = [
+        COST_UNEXPECTED_MESSAGE,
+        COST_CORRUPTED_MESSAGE,
+        COST_NETWORK_ERROR,
+        COST_REQUEST_TIMED_OUT,
+        COST_INVALID_SIGNATURE,
+        COST_WRONG_PARA,
+        COST_UNNEEDED_COLLATOR,
+        COST_REPORT_BAD,
+        BENEFIT_NOTIFY_GOOD,
+    ];
+
+    pub enum FitnessEvent {
+        Unexpected,
+        Corrupted,
+        NetworkError,
+        Timeout,
+        SigError,
+        ParaError,
+        Superfluous,
+        ReportBad,
+        NotifyGood,
+    }
+
+    impl From<FitnessEvent> for usize {
+        fn from(input: FitnessEvent) -> usize {
+            match input {
+                FitnessEvent::Unexpected => 0,
+                FitnessEvent::Corrupted => 1,
+                FitnessEvent::NetworkError => 2,
+                FitnessEvent::Timeout => 3,
+                FitnessEvent::SigError => 4,
+                FitnessEvent::ParaError => 5,
+                FitnessEvent::Superfluous => 6,
+                FitnessEvent::ReportBad => 7,
+                FitnessEvent::NotifyGood => 8,
+            }
+        }
+    }
```

Comment:
FWIW, we used to have a `repr(usize)` enum, but opted for a more type-safe API
https://github.com/paritytech/polkadot/pull/2420/files